File: rclpy/rclpy/qos_event.py
# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from enum import IntEnum
from typing import Callable
from typing import List
from typing import Optional

import rclpy
from rclpy.callback_groups import CallbackGroup
from rclpy.handle import Handle
from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
from rclpy.logging import get_logger
from rclpy.qos import qos_policy_name_from_kind
from rclpy.waitable import NumberOfEntities
from rclpy.waitable import Waitable


QoSPublisherEventType = _rclpy.rcl_publisher_event_type_t
QoSSubscriptionEventType = _rclpy.rcl_subscription_event_type_t


# Payload type for Subscription Deadline callback.
QoSRequestedDeadlineMissedInfo = _rclpy.rmw_requested_deadline_missed_status_t

# Payload type for Subscription Liveliness callback.
QoSLivelinessChangedInfo = _rclpy.rmw_liveliness_changed_status_t

# Payload type for Subscription Message Lost callback.
QoSMessageLostInfo = _rclpy.rmw_message_lost_status_t

# Payload type for Subscription Incompatible QoS callback.
QoSRequestedIncompatibleQoSInfo = _rclpy.rmw_requested_qos_incompatible_event_status_t

# Payload type for Publisher Deadline callback.
QoSOfferedDeadlineMissedInfo = _rclpy.rmw_offered_deadline_missed_status_t

# Payload type for Publisher Liveliness callback.
QoSLivelinessLostInfo = _rclpy.rmw_liveliness_lost_status_t

"""
Payload type for Publisher Incompatible QoS callback.

Mirrors rmw_offered_incompatible_qos_status_t from rmw/types.h
"""
QoSOfferedIncompatibleQoSInfo = QoSRequestedIncompatibleQoSInfo


"""Raised when registering a callback for an event type that is not supported."""
UnsupportedEventTypeError = _rclpy.UnsupportedEventTypeError


class QoSEventHandler(Waitable):
    """Waitable type to handle QoS events."""

    def __init__(
        self,
        *,
        callback_group: CallbackGroup,
        callback: Callable,
        event_type: IntEnum,
        parent_handle: Handle,
    ):
        # Waitable init adds self to callback_group
        super().__init__(callback_group)
        self.event_type = event_type
        self.callback = callback

        with parent_handle as parent_capsule:
            self.__event = _rclpy.QoSEvent(parent_capsule, event_type)
        self._ready_to_take_data = False
        self._event_index = None

    # Start Waitable API
    def is_ready(self, wait_set):
        """Return True if entities are ready in the wait set."""
        if self._event_index is None:
            return False
        if _rclpy.rclpy_wait_set_is_ready('event', wait_set, self._event_index):
            self._ready_to_take_data = True
        return self._ready_to_take_data

    def take_data(self):
        """Take stuff from lower level so the wait set doesn't immediately wake again."""
        if self._ready_to_take_data:
            self._ready_to_take_data = False
            with self.__event:
                return self.__event.take_event()
        return None

    async def execute(self, taken_data):
        """Execute work after data has been taken from a ready wait set."""
        if not taken_data:
            return
        await rclpy.executors.await_or_execute(self.callback, taken_data)

    def get_num_entities(self):
        """Return number of each type of entity used."""
        return NumberOfEntities(num_events=1)

    def add_to_wait_set(self, wait_set):
        """Add entites to wait set."""
        with self.__event:
            self._event_index = _rclpy.rclpy_wait_set_add_event(wait_set, self.__event)

    def __enter__(self):
        """Mark event as in-use to prevent destruction while waiting on it."""
        self.__event.__enter__()

    def __exit__(self, t, v, tb):
        """Mark event as not-in-use to allow destruction after waiting on it."""
        self.__event.__exit__(t, v, tb)

    def destroy(self):
        self.__event.destroy_when_not_in_use()


class SubscriptionEventCallbacks:
    """Container to provide middleware event callbacks for a Subscription."""

    def __init__(
        self,
        *,
        deadline: Optional[Callable[[QoSRequestedDeadlineMissedInfo], None]] = None,
        incompatible_qos: Optional[Callable[[QoSRequestedIncompatibleQoSInfo], None]] = None,
        liveliness: Optional[Callable[[QoSLivelinessChangedInfo], None]] = None,
        message_lost: Optional[Callable[[QoSMessageLostInfo], None]] = None,
        use_default_callbacks: bool = True,
    ) -> None:
        """
        Create a SubscriptionEventCallbacks container.

        :param deadline: A user-defined callback that is called when a topic misses our
            requested Deadline.
        :param incompatible_qos: A user-defined callback that is called when a Publisher
            with incompatible QoS policies is discovered on subscribed topic.
        :param liveliness: A user-defined callback that is called when the Liveliness of
            a Publisher on subscribed topic changes.
        :param message_lost: A user-defined callback that is called when a messages is lost.
        :param use_default_callbacks: Whether or not to use default callbacks when the user
            doesn't supply one
        """
        self.deadline = deadline
        self.incompatible_qos = incompatible_qos
        self.liveliness = liveliness
        self.message_lost = message_lost
        self.use_default_callbacks = use_default_callbacks

    def create_event_handlers(
        self, callback_group: CallbackGroup, subscription_handle: Handle, topic_name: str,
    ) -> List[QoSEventHandler]:
        with subscription_handle as subscription_capsule:
            logger = get_logger(_rclpy.rclpy_get_subscription_logger_name(subscription_capsule))

        event_handlers = []
        if self.deadline:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.deadline,
                event_type=QoSSubscriptionEventType.RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED,
                parent_handle=subscription_handle))

        if self.incompatible_qos:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.incompatible_qos,
                event_type=QoSSubscriptionEventType.RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS,
                parent_handle=subscription_handle))
        elif self.use_default_callbacks:
            # Register default callback when not specified
            try:
                def _default_incompatible_qos_callback(event):
                    policy_name = qos_policy_name_from_kind(event.last_policy_kind)
                    logger.warn(
                        "New publisher discovered on topic '{}', offering incompatible QoS. "
                        'No messages will be received from it. '
                        'Last incompatible policy: {}'.format(topic_name, policy_name))

                event_type = QoSSubscriptionEventType.RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS
                event_handlers.append(QoSEventHandler(
                    callback_group=callback_group,
                    callback=_default_incompatible_qos_callback,
                    event_type=event_type,
                    parent_handle=subscription_handle))

            except UnsupportedEventTypeError:
                pass

        if self.liveliness:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.liveliness,
                event_type=QoSSubscriptionEventType.RCL_SUBSCRIPTION_LIVELINESS_CHANGED,
                parent_handle=subscription_handle))

        if self.message_lost:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.message_lost,
                event_type=QoSSubscriptionEventType.RCL_SUBSCRIPTION_MESSAGE_LOST,
                parent_handle=subscription_handle))

        return event_handlers


class PublisherEventCallbacks:
    """Container to provide middleware event callbacks for a Publisher."""

    def __init__(
        self,
        *,
        deadline: Optional[Callable[[QoSOfferedDeadlineMissedInfo], None]] = None,
        liveliness: Optional[Callable[[QoSLivelinessLostInfo], None]] = None,
        incompatible_qos: Optional[Callable[[QoSRequestedIncompatibleQoSInfo], None]] = None,
        use_default_callbacks: bool = True,
    ) -> None:
        """
        Create and return a PublisherEventCallbacks container.

        :param deadline: A user-defined callback that is called when the Publisher misses
            its offered Deadline.
        :param liveliness: A user-defined callback that is called when this Publisher
            fails to signal its Liveliness and is reported as not-alive.
        :param incompatible_qos: A user-defined callback that is called when a Subscription
            with incompatible QoS policies is discovered on subscribed topic.
        :param use_default_callbacks: Whether or not to use default callbacks when the user
            doesn't supply one
        """
        self.deadline = deadline
        self.liveliness = liveliness
        self.incompatible_qos = incompatible_qos
        self.use_default_callbacks = use_default_callbacks

    def create_event_handlers(
        self, callback_group: CallbackGroup, publisher_handle: Handle, topic_name: str,
    ) -> List[QoSEventHandler]:
        with publisher_handle as publisher_capsule:
            logger = get_logger(_rclpy.rclpy_get_publisher_logger_name(publisher_capsule))

        event_handlers = []
        if self.deadline:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.deadline,
                event_type=QoSPublisherEventType.RCL_PUBLISHER_OFFERED_DEADLINE_MISSED,
                parent_handle=publisher_handle))

        if self.liveliness:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.liveliness,
                event_type=QoSPublisherEventType.RCL_PUBLISHER_LIVELINESS_LOST,
                parent_handle=publisher_handle))

        if self.incompatible_qos:
            event_handlers.append(QoSEventHandler(
                callback_group=callback_group,
                callback=self.incompatible_qos,
                event_type=QoSPublisherEventType.RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS,
                parent_handle=publisher_handle))
        elif self.use_default_callbacks:
            # Register default callback when not specified
            try:
                def _default_incompatible_qos_callback(event):
                    policy_name = qos_policy_name_from_kind(event.last_policy_kind)
                    logger.warn(
                        "New subscription discovered on topic '{}', requesting incompatible QoS. "
                        'No messages will be sent to it. '
                        'Last incompatible policy: {}'.format(topic_name, policy_name))

                event_handlers.append(QoSEventHandler(
                    callback_group=callback_group,
                    callback=_default_incompatible_qos_callback,
                    event_type=QoSPublisherEventType.RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS,
                    parent_handle=publisher_handle))

            except UnsupportedEventTypeError:
                pass

        return event_handlers

File: rclpy/src/rclpy/qos_event.cpp
// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <pybind11/pybind11.h>

#include <rcl/error_handling.h>
#include <rcl/event.h>
#include <rmw/incompatible_qos_events_statuses.h>

#include <memory>
#include <stdexcept>
#include <utility>

#include "exceptions.hpp"
#include "qos_event.hpp"

namespace rclpy
{
static
std::shared_ptr<rcl_event_t>
create_zero_initialized_event()
{
  auto event = std::shared_ptr<rcl_event_t>(
    new rcl_event_t,
    [](rcl_event_t * event)
    {
      rcl_ret_t ret = rcl_event_fini(event);
      if (RCL_RET_OK != ret) {
        int stack_level = 1;
        PyErr_WarnFormat(
          PyExc_RuntimeWarning, stack_level,
          "failed to fini event: %s",
          rcl_get_error_string().str);
        rcl_reset_error();
      }
      delete event;
    });

  *event = rcl_get_zero_initialized_event();
  return event;
}

void
QoSEvent::destroy()
{
  rcl_event_.reset();
  std::visit([](auto & t) {t.destroy();}, grandparent_);
}

QoSEvent::QoSEvent(
  rclpy::Subscription & subscription, rcl_subscription_event_type_t event_type)
: event_type_(event_type), grandparent_(subscription)
{
  // Create a subscription event
  rcl_event_ = create_zero_initialized_event();

  rcl_ret_t ret = rcl_subscription_event_init(
    rcl_event_.get(),
    std::get<Subscription>(grandparent_).rcl_ptr(), event_type);
  if (RCL_RET_BAD_ALLOC == ret) {
    rcl_reset_error();
    throw std::bad_alloc();
  }
  if (RCL_RET_UNSUPPORTED == ret) {
    throw UnsupportedEventTypeError("subscription event is unsupported");
  }
  if (RCL_RET_OK != ret) {
    throw RCLError("failed to create subscription event");
  }
}

QoSEvent::QoSEvent(
  rclpy::Publisher & publisher, rcl_publisher_event_type_t event_type)
: event_type_(event_type), grandparent_(publisher)
{
  // Create a publisher event
  rcl_event_ = create_zero_initialized_event();

  rcl_ret_t ret = rcl_publisher_event_init(
    rcl_event_.get(),
    std::get<Publisher>(grandparent_).rcl_ptr(), event_type);
  if (RCL_RET_BAD_ALLOC == ret) {
    rcl_reset_error();
    throw std::bad_alloc();
  }
  if (RCL_RET_UNSUPPORTED == ret) {
    throw UnsupportedEventTypeError("publisher event is unsupported");
  }
  if (RCL_RET_OK != ret) {
    throw RCLError("failed to create publisher event");
  }
}

typedef union qos_event_callback_data {
  // Subscription events
  rmw_requested_deadline_missed_status_t requested_deadline_missed;
  rmw_liveliness_changed_status_t liveliness_changed;
  rmw_message_lost_status_t message_lost;
  rmw_requested_qos_incompatible_event_status_t requested_incompatible_qos;
  // Publisher events
  rmw_offered_deadline_missed_status_t offered_deadline_missed;
  rmw_liveliness_lost_status_t liveliness_lost;
  rmw_offered_qos_incompatible_event_status_t offered_incompatible_qos;
} qos_event_callback_data_t;

py::object
QoSEvent::take_event()
{
  qos_event_callback_data_t data;
  rcl_ret_t ret = rcl_take_event(rcl_event_.get(), &data);
  if (RCL_RET_BAD_ALLOC == ret) {
    rcl_reset_error();
    throw std::bad_alloc();
  }
  if (RCL_RET_EVENT_TAKE_FAILED == ret) {
    return py::none();
  }
  if (RCL_RET_OK != ret) {
    throw RCLError("failed to take event");
  }

  if (auto sub_type = std::get_if<rcl_subscription_event_type_t>(&event_type_)) {
    switch (*sub_type) {
      case RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED:
        return py::cast(data.requested_deadline_missed);
      case RCL_SUBSCRIPTION_LIVELINESS_CHANGED:
        return py::cast(data.liveliness_changed);
      case RCL_SUBSCRIPTION_MESSAGE_LOST:
        return py::cast(data.message_lost);
      case RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS:
        return py::cast(data.requested_incompatible_qos);
      default:
        // suggests a misalignment between C and Python interfaces
        throw py::value_error("event type for subscriptions not understood");
    }
  } else if (auto pub_type = std::get_if<rcl_publisher_event_type_t>(&event_type_)) {
    switch (*pub_type) {
      case RCL_PUBLISHER_OFFERED_DEADLINE_MISSED:
        return py::cast(data.offered_deadline_missed);
      case RCL_PUBLISHER_LIVELINESS_LOST:
        return py::cast(data.liveliness_lost);
      case RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS:
        return py::cast(data.offered_incompatible_qos);
      default:
        // suggests a misalignment between C and Python interfaces
        throw py::value_error("event type for publishers not understood");
    }
  }
  throw std::runtime_error("cannot take event that is neither a publisher or a subscription event");
}

void
define_qos_event(py::module module)
{
  py::class_<QoSEvent, Destroyable, std::shared_ptr<QoSEvent>>(module, "QoSEvent")
  .def(py::init<rclpy::Subscription &, rcl_subscription_event_type_t>())
  .def(py::init<rclpy::Publisher &, rcl_publisher_event_type_t>())
  .def_property_readonly(
    "pointer", [](const QoSEvent & event) {
      return reinterpret_cast<size_t>(event.rcl_ptr());
    },
    "Get the address of the entity as an integer")
  .def(
    "take_event", &QoSEvent::take_event,
    "Get pending data from a ready QoS event");

  py::enum_<rcl_subscription_event_type_t>(module, "rcl_subscription_event_type_t")
  .value("RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED", RCL_SUBSCRIPTION_REQUESTED_DEADLINE_MISSED)
  .value("RCL_SUBSCRIPTION_LIVELINESS_CHANGED", RCL_SUBSCRIPTION_LIVELINESS_CHANGED)
  .value("RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS", RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS)
  .value("RCL_SUBSCRIPTION_MESSAGE_LOST", RCL_SUBSCRIPTION_MESSAGE_LOST);

  py::enum_<rcl_publisher_event_type_t>(module, "rcl_publisher_event_type_t")
  .value("RCL_PUBLISHER_OFFERED_DEADLINE_MISSED", RCL_PUBLISHER_OFFERED_DEADLINE_MISSED)
  .value("RCL_PUBLISHER_LIVELINESS_LOST", RCL_PUBLISHER_LIVELINESS_LOST)
  .value("RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS", RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS);

  py::class_<rmw_requested_deadline_missed_status_t>(
    module, "rmw_requested_deadline_missed_status_t")
  .def(py::init<>())
  .def_readonly("total_count", &rmw_requested_deadline_missed_status_t::total_count)
  .def_readonly("total_count_change", &rmw_requested_deadline_missed_status_t::total_count_change);

  py::class_<rmw_liveliness_changed_status_t>(module, "rmw_liveliness_changed_status_t")
  .def(py::init<>())
  .def_readonly("alive_count", &rmw_liveliness_changed_status_t::alive_count)
  .def_readonly("not_alive_count", &rmw_liveliness_changed_status_t::not_alive_count)
  .def_readonly("alive_count_change", &rmw_liveliness_changed_status_t::alive_count_change)
  .def_readonly("not_alive_count_change", &rmw_liveliness_changed_status_t::not_alive_count_change);

  py::class_<rmw_message_lost_status_t>(module, "rmw_message_lost_status_t")
  .def(py::init<>())
  .def_readonly("total_count", &rmw_message_lost_status_t::total_count)
  .def_readonly("total_count_change", &rmw_message_lost_status_t::total_count_change);

  py::class_<rmw_requested_qos_incompatible_event_status_t>(
    module, "rmw_requested_qos_incompatible_event_status_t")
  .def(py::init<>())
  .def_readonly("total_count", &rmw_requested_qos_incompatible_event_status_t::total_count)
  .def_readonly(
    "total_count_change", &rmw_requested_qos_incompatible_event_status_t::total_count_change)
  .def_readonly(
    "last_policy_kind", &rmw_requested_qos_incompatible_event_status_t::last_policy_kind);

  py::class_<rmw_offered_deadline_missed_status_t>(module, "rmw_offered_deadline_missed_status_t")
  .def(py::init<>())
  .def_readonly("total_count", &rmw_offered_deadline_missed_status_t::total_count)
  .def_readonly("total_count_change", &rmw_offered_deadline_missed_status_t::total_count_change);

  py::class_<rmw_liveliness_lost_status_t>(module, "rmw_liveliness_lost_status_t")
  .def(py::init<>())
  .def_readonly("total_count", &rmw_liveliness_lost_status_t::total_count)
  .def_readonly("total_count_change", &rmw_liveliness_lost_status_t::total_count_change);

  py::enum_<rmw_qos_policy_kind_t>(module, "rmw_qos_policy_kind_t")
  .value("RMW_QOS_POLICY_INVALID", RMW_QOS_POLICY_INVALID)
  .value("RMW_QOS_POLICY_DURABILITY", RMW_QOS_POLICY_DURABILITY)
  .value("RMW_QOS_POLICY_DEADLINE", RMW_QOS_POLICY_DEADLINE)
  .value("RMW_QOS_POLICY_LIVELINESS", RMW_QOS_POLICY_LIVELINESS)
  .value("RMW_QOS_POLICY_RELIABILITY", RMW_QOS_POLICY_RELIABILITY)
  .value("RMW_QOS_POLICY_HISTORY", RMW_QOS_POLICY_HISTORY)
  .value("RMW_QOS_POLICY_LIFESPAN", RMW_QOS_POLICY_LIFESPAN)
  .value("RMW_QOS_POLICY_DEPTH", RMW_QOS_POLICY_DEPTH)
  .value("RMW_QOS_POLICY_LIVELINESS_LEASE_DURATION", RMW_QOS_POLICY_LIVELINESS_LEASE_DURATION)
  .value(
    "RMW_QOS_POLICY_AVOID_ROS_NAMESPACE_CONVENTIONS",
    RMW_QOS_POLICY_AVOID_ROS_NAMESPACE_CONVENTIONS);
}
}  // namespace rclpy
